nilearn/plotting/surf_plotting.py

```diff
@@ -849,17 +849,25 @@ def plot_img_on_surf(stat_map, surf_mesh='fsaverage5', mask_img=None,
     for ax in axes.flatten():
         # We increase this value to better position the camera of the
         # 3D projection plot. The default value makes meshes look too small.
-        ax.dist = 6
+        ax.dist = 7 
 
     if colorbar:
         sm = _colorbar_from_array(image.get_data(stat_map),
                                   vmax, threshold, kwargs,
                                   cmap=get_cmap(cmap))
-
-        cbar_ax = fig.add_subplot(32, 1, 32)
+       
+        # if figure has two rows, change colorbar position for less excess space
+        if (axes.shape[0] == 2):
+            cbar_ax = fig.add_axes([0.265,0.10,0.5,0.04]) 
+        else:
+            cbar_ax = fig.add_axes([0.265,0.17,0.5,0.04]) 
```

Comment:
Are there only these two possibilities? 
I might be missing something but I don't see why you couldn't have more than 2 rows :thinking:

Reply:
I guess I was only focusing on the two rows, but you are right. My new commit should handle three rows and beyond. 

![download (8)](https://user-images.githubusercontent.com/8116564/117354345-d8b44980-ae76-11eb-8f87-fb2bb346bf53.png)

Reply:
maybe using a gridspec that takes into account the colorbar instead of not taking it into account in the layout and adding it on top afterwards could help avoid these special cases and fine-tuning? 
maybe not, these things are tricky especially with 3d axes.

Reply:
I have the impression that unless it is passes a figure as well as the axes, `plot_surf` messes up the figure size:

https://github.com/nilearn/nilearn/blob/f76dabd6baf91db487cec606a7d0ea46effb50a2/nilearn/plotting/surf_plotting.py#L220

I think the figure size will always be the same whatever we set here 

https://github.com/nilearn/nilearn/blob/f76dabd6baf91db487cec606a7d0ea46effb50a2/nilearn/plotting/surf_plotting.py#L834

which probably doesn't help

Reply:
(try replacing figsize in the line above with whatever pair of values it doesn't affect the figure)

Reply:
I realized that we don't currently check views and hemispheres at all. 
So, you can do crazy things like:

```python
from nilearn.datasets import fetch_neurovault_motor_task
from nilearn.plotting import plot_img_on_surf

motor_images = fetch_neurovault_motor_task()
stat_img = motor_images.images[0]
plot_img_on_surf(stat_img,
                 views=['lateral','lateral','lateral'],
                 hemispheres=['left','left','left'],
                 colorbar=True)
```
which look terrible:

![plot_img](https://user-images.githubusercontent.com/2639645/117420630-0f916a80-af1e-11eb-9a03-584d8ef652a6.png)

Should we constrain these ? (might make the plot setup easier...)

Reply:
I think it's ok to allow it -- if someone asks for the same view twice they get it twice which seems reasonable, unsurprising behavior. of course it would be best if they didn't get cut off but (i) I think the problem you see here is related to the bug I pointed out above rather to the fact that the same view is used several times (ii) if the plot doesn't look as good (but still doesn't result in an error or really incorrect results) in a corner case that no one will use it's not a big problem

Reply:
for example if you were to apply this patch (bg color, hspace and wspace is just to help see axis boundaries):

```diff
diff --git a/nilearn/plotting/surf_plotting.py b/nilearn/plotting/surf_plotting.py
index 8baa0b9d8..6b8d7822a 100644
--- a/nilearn/plotting/surf_plotting.py
+++ b/nilearn/plotting/surf_plotting.py
@@ -2,9 +2,12 @@
 Functions for surface visualization.
 Only matplotlib is required.
 """
+import itertools
+
 import matplotlib.pyplot as plt
 import numpy as np
 
+from matplotlib import gridspec
 from matplotlib.colorbar import make_axes
 from matplotlib.cm import ScalarMappable, get_cmap
 from matplotlib.colors import Normalize, LinearSegmentedColormap
@@ -227,7 +230,7 @@ def plot_surf(surf_mesh, surf_map=None, bg_map=None,
     else:
         if figure is None:
             figure = axes.get_figure()
-            figure.set_size_inches(*figsize)
+            # figure.set_size_inches(*figsize)
         axes.set_xlim(*limits)
         axes.set_ylim(*limits)
     axes.view_init(elev=elev, azim=azim)
@@ -769,7 +772,7 @@ def plot_img_on_surf(stat_map, surf_mesh='fsaverage5', mask_img=None,
                      views=['lateral', 'medial'],
                      output_file=None, title=None, colorbar=True,
                      vmax=None, threshold=None,
-                     cmap='cold_hot', aspect_ratio=1.4, **kwargs):
+                     cmap='cold_hot', **kwargs):
     """Convenience function to plot multiple views of plot_surf_stat_map
     in a single figure. It projects stat_map into meshes and plots views of
     left and right hemispheres. The *views* argument defines the views
@@ -874,45 +877,41 @@ def plot_img_on_surf(stat_map, surf_mesh='fsaverage5', mask_img=None,
                              mask_img=mask_img)
     }
 
-    figsize = plt.figaspect(len(modes) / (aspect_ratio * len(hemispheres)))
-    fig, axes = plt.subplots(nrows=len(modes),
-                             ncols=len(hemis),
-                             figsize=figsize,
-                             subplot_kw={'projection': '3d'})
-
-    axes = np.atleast_2d(axes)
-
-    if len(hemis) == 1:
-        axes = axes.T
-
-    for index_mode, mode in enumerate(modes):
-        for index_hemi, hemi in enumerate(hemis):
-            bg_map = surf_mesh['sulc_%s' % hemi]
-            plot_surf_stat_map(surf[hemi], texture[hemi],
-                               view=mode, hemi=hemi,
-                               bg_map=bg_map,
-                               axes=axes[index_mode, index_hemi],
-                               colorbar=False,  # Colorbar created externally.
-                               vmax=vmax,
-                               threshold=threshold,
-                               cmap=cmap,
-                               **kwargs)
-
-    for ax in axes.flatten():
+    cbar_h = .1
+    w, h = plt.figaspect((len(modes) + cbar_h) / len(hemispheres))
+    fig = plt.figure(figsize=(w, h))
+    grid = gridspec.GridSpec(
+        len(modes) + 1, len(hemis),
+        height_ratios=[1.] * len(modes) + [cbar_h], hspace=0.05, wspace=0.05)
+    axes = []
+    for i, (mode, hemi) in enumerate(itertools.product(modes, hemis)):
+        bg_map = surf_mesh['sulc_%s' % hemi]
+        ax = fig.add_subplot(grid[i], projection="3d")
+        axes.append(ax)
+        plot_surf_stat_map(surf[hemi], texture[hemi],
+                           view=mode, hemi=hemi,
+                           bg_map=bg_map,
+                           axes=ax,
+                           colorbar=False,  # Colorbar created externally.
+                           vmax=vmax,
+                           threshold=threshold,
+                           cmap=cmap,
+                           **kwargs)
+        ax.set_facecolor("#e0e0e0")
         # We increase this value to better position the camera of the
         # 3D projection plot. The default value makes meshes look too small.
-        ax.dist = 6
+        ax.dist = 7
 
     if colorbar:
         sm = _colorbar_from_array(image.get_data(stat_map),
                                   vmax, threshold, kwargs,
                                   cmap=get_cmap(cmap))
 
-        cbar_ax = fig.add_subplot(32, 1, 32)
+        cbar_grid = gridspec.GridSpecFromSubplotSpec(1, 3, grid[-1, :])
+        cbar_ax = fig.add_subplot(cbar_grid[1])
+        axes.append(cbar_ax)
         fig.colorbar(sm, cax=cbar_ax, orientation='horizontal')
 
-    fig.subplots_adjust(wspace=-0.02, hspace=0.0)
-
     if title is not None:
         fig.suptitle(title)
 
```

the output of the script you posted above would be

![fignico](https://user-images.githubusercontent.com/9196501/117455445-313a2400-af15-11eb-8b18-0df7ecbf54d2.png)

Reply:
and a more realistic one:
![fig](https://user-images.githubusercontent.com/9196501/117456361-1b792e80-af16-11eb-8776-f7ab7db3714d.png)

Reply:
Wow, that looks excellent @jeromedockes ! Should I make those changes, and then commit those to this pull request? Or what's the best way forward?

Reply:
maybe the easiest way to apply this patch is for me to open a PR on your `plot_fix` branch. you can then review and merge it into this branch (from the github interface), and keep pushing to the same branch to update this PR with any adjustments you deem necessary, and when we are finally satisfied we merge your branch into upstream.
does that sound reasonable

Reply:
Sounds good!

Reply:
ok great let's do that
I opened a PR, it looks like a large diff but this is just because your branch is a bit behind master. if you merge master into your branch first it will be easier to check!